pkg/stanza/fileconsumer/file_threadpool.go

```diff
@@ -0,0 +1,198 @@
+
+	// Any new files that appear should be consumed entirely
+	m.readerFactory.fromBeginning = true
+	m.syncLastPollFilesConcurrent(ctx)
+}
+
+func (m *Manager) worker(ctx context.Context) {
+	defer m.workerWg.Done()
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case chanData, ok := <-m.readerChan:
+			if !ok {
+				return
+			}
+			r, fp := chanData.reader, chanData.fp
+			if !m.readToEnd(ctx, r) {
+				// Save off any files that were not fully read or if deleteAfterRead is disabled
+				m.knownFilesLock.Lock()
+				m.knownFiles = append(m.knownFiles, r)
+				m.knownFilesLock.Unlock()
+			}
+			m.removePath(fp)
+		}
+	}
+}
+
+func (m *Manager) makeReaderConcurrent(filePath string) (*reader, *fingerprint.Fingerprint) {
+	fp, file := m.makeFingerprint(filePath)
+	if fp == nil {
+		return nil, nil
+	}
+
+	// check if the current file is already being consumed
+	if m.isCurrentlyConsuming(fp) {
+		if err := file.Close(); err != nil {
+			m.Errorf("problem closing file", "file", file.Name())
+		}
+		return nil, nil
+	}
+
+	// Exclude any empty fingerprints or duplicate fingerprints to avoid doubling up on copy-truncate files
+	if m.checkDuplicates(fp) {
+		if err := file.Close(); err != nil {
+			m.Errorf("problem closing file", "file", file.Name())
+		}
+		return nil, nil
+	}
+	m.currentFps = append(m.currentFps, fp)
+
+	reader, err := m.newReaderConcurrent(file, fp)
+	if err != nil {
+		m.Errorw("Failed to create reader", zap.Error(err))
+		return nil, nil
+	}
+	return reader, fp
+}
+
+func (m *Manager) consumeConcurrent(ctx context.Context, paths []string) {
+	m.clearOldReadersConcurrent(ctx)
+	for _, path := range paths {
+		reader, fp := m.makeReaderConcurrent(path)
+		if reader != nil {
+			// add path and fingerprint as it's not consuming
+			m.trieLock.Lock()
+			m.trie.Put(fp.FirstBytes)
+			m.trieLock.Unlock()
+			m.readerChan <- readerWrapper{reader: reader, fp: fp}
+		}
+	}
+}
+
+func (m *Manager) isCurrentlyConsuming(fp *fingerprint.Fingerprint) bool {
+	m.trieLock.RLock()
+	defer m.trieLock.RUnlock()
+	return m.trie.HasKey(fp.FirstBytes)
+}
+
+func (m *Manager) removePath(fp *fingerprint.Fingerprint) {
+	m.trieLock.Lock()
+	defer m.trieLock.Unlock()
+	m.trie.Delete(fp.FirstBytes)
+}
+
+func (m *Manager) clearOldReadersConcurrent(ctx context.Context) {
+	m.knownFilesLock.Lock()
+	defer m.knownFilesLock.Unlock()
+	// Clear out old readers. They are sorted such that they are oldest first,
+	// so we can just find the first reader whose poll cycle is less than our
+	// limit i.e. last 3 cycles, and keep every reader after that
+	oldReaders := make([]*reader, 0)
+	i := 0
+	for ; i < len(m.knownFiles); i++ {
+		reader := m.knownFiles[i]
+		if reader.generation >= 3 {
+			oldReaders = append(oldReaders, reader)
+		} else {
+			break
+		}
+	}
+	m.knownFiles = m.knownFiles[i:]
+
+	var lostWG sync.WaitGroup
+	for _, r := range oldReaders {
+		lostWG.Add(1)
+		go func(r *reader) {
+			defer lostWG.Done()
+			r.ReadToEnd(ctx)
```

Comment:
I think this is ok because `clearOldReadersConcurrent` is blocking, but if we want to be truly asynchronous, where a new poll may find the same file and begin reading it, then this would potentially result in reading the same content multiple times (once here, and once from a handle found in the next poll).